src/ipips/ipip-0412.md

```diff
@@ -0,0 +1,208 @@
+   in the CAR response, the Trustless Gateway could have a server-side
+   configuration that specifies the default order. However, this approach would
+   limit the flexibility for clients, requiring them to have prior knowledge
+   about order supported by each gateway.
+
+2. Fixed Block Order: Another option was to enforce a fixed block order in the
+   CAR responses. However, this approach would not cater to the varying needs
+   and preferences of different clients and use cases, and is not backward
+   compatible with the existing Trustless Gateways which return CAR responses
+   with Weak `Etag` and unspecified block order.
+
+3. Separate `X-` HTTP Header: Introduction of a separate HTTP reader was
+   rejected because we try to use HTTP semantics where possible, and gateways
+   already use HTTP content type negotiation for CAR `version` and reusing it
+   saves a few bytes in each round-trip. Also, :cite[rfc6648] advises against
+   use of `X-` and similar constructs in new protocols.
+
+4. The decision to not implement a single preset pack with predefined behavior,
+   instead of separate parameters for order and duplicates (dups), was driven
+   by considerations of ambiguity and potential future problems when adding
+   more determinism to responses. For instance, if we were to include a new
+   behavior like `foo=y|n` alongside an existing preset like `pack=orderdfs+dupsy`,
+   it would either necessitate the addition of a separate parameter or impose
+   the adoption of a new version of every preset (e.g., `orderdfs-dupsy+fooy` and
+   `orderdfs+dupsy+foon`). Maintaining and deploying such changes across a
+   decentralized ecosystem, where gateways may operate on different software,
+   becomes more complex. In contrast, utilizing separate parameters for each
+   behavior enables easier maintenance and deployment in a decentralized
+   ecosystem with varying gateway software.
+
+The proposed solution of negotiating the block order through headers is
+future-proof, allows for flexibility, interoperability, and customization while
+maintaining compatibility with existing implementations.
+
+## Test fixtures
+
+Implementation compliance can be determined by testing the negotiation process
+between clients and Trustless Gateways using various combinations of `order` and
+`dups` parameters.
+
+Relevant tests were added to
+[gateway-conformance](https://github.com/ipfs/gateway-conformance) test suite
+in [#87](https://github.com/ipfs/gateway-conformance/pull/87).
+
+<!-- TODO after conformance release is tagged
+A detailed list of compliance checks for `order` and `dupes` can be found in
+[`v0.TODO.0/trustless_gateway_car_test.go`](https://github.com/ipfs/gateway-conformance/blob/v0.TODO.0/tests/trustless_gateway_car_test.go) or later.
+-->
+
+Below are CIDs, CARs, and short summary of each fixture.
+
+TODO:
+1. a CAR with blocks for a small file in DFS order
+2. a CAR with blocks for a small file with one block appearing twice
```

Comment:
- [ ] This needs to be updated after https://github.com/ipfs/gateway-conformance/issues/77 is merged.

Not doing this yet to make sure IPIP includes the same CIDs as conformance test suite.